### PR TITLE
refactor stats API

### PR DESCRIPTION
**Pull request by ylwu-amzn**
Signed-off-by: Yaliang Wu <ylwu@amazon.com>

### Description
Refactor  stats API. 

Main changes:
1. Add new cluster level stats: `ml_model_count`, `ml_task_index_status`, `ml_model_index_status`
2. Split stats into 4 levels: cluster, node, algorithm and action.
3. Change stats REST API interface. User can query specific stats on each level and for different stats.
For example, if user want to query only kmeans stats, the request and response will be.

```
GET /_plugins/_ml/stats
{
  "target_stat_levels": ["algorithm"],
  "algorithms": ["kmeans"]
}

# Response
{
  "nodes" : {
    "9smbhzFHRgCOwuzp48j3xw" : {
      "algorithms" : { }
    },
    "GHwAwzBxQkifTPzs5lt3gQ" : {
      "algorithms" : {
        "kmeans" : {
          "train" : {
            "ml_action_request_count" : 1
          }
        }
      }
    },
    "xa4IaPIMQ166q5MZyUrgFg" : {
      "algorithms" : { }
    },
    "xbgmGZ1jRUi_JTfmgjkstg" : {
      "algorithms" : { }
    },
    "niuqXoKCTAq2pNmm8cUi7g" : {
      "algorithms" : { }
    },
    "PBCe4MIqSlWtjGIxKmCzFw" : {
      "algorithms" : {
        "kmeans" : {
          "train_predict" : {
            "ml_action_request_count" : 1
          }
        }
      }
    },
    "qP6kU6DbQCa1T-hByDqw_A" : {
      "algorithms" : { }
    }
  }
}
```

User can query stats on different levels and filter by combined conditions. For example, user can query cluster and algorithm stats like this
```
GET /_plugins/_ml/stats
{
  "target_stat_levels": [
    "cluster",
    "algorithm"
  ],
  "algorithms": [
    "kmeans","anomaly_localization"
  ],
  "node_level_stats": [
    "ml_node_total_request_count","ml_node_total_failure_count"
  ],
  "cluster_level_stats": [
    "ml_model_count"
  ]
}
```
Response
```
{
  "ml_model_count" : 1,
  "nodes" : {
    "9smbhzFHRgCOwuzp48j3xw" : {
      "algorithms" : { }
    },
    "xa4IaPIMQ166q5MZyUrgFg" : {
      "algorithms" : {
        "anomaly_localization" : {
          "execute" : {
            "ml_action_request_count" : 2
          }
        }
      }
    },
    "GHwAwzBxQkifTPzs5lt3gQ" : {
      "algorithms" : {
        "kmeans" : {
          "train" : {
            "ml_action_request_count" : 1
          }
        },
        "anomaly_localization" : {
          "execute" : {
            "ml_action_request_count" : 4
          }
        }
      }
    },
    "PBCe4MIqSlWtjGIxKmCzFw" : {
      "algorithms" : {
        "kmeans" : {
          "train_predict" : {
            "ml_action_request_count" : 1
          }
        }
      }
    },
    "xbgmGZ1jRUi_JTfmgjkstg" : {
      "algorithms" : { }
    },
    "qP6kU6DbQCa1T-hByDqw_A" : {
      "algorithms" : {
        "kmeans" : {
          "train" : {
            "ml_action_request_count" : 2
          }
        },
        "anomaly_localization" : {
          "execute" : {
            "ml_action_request_count" : 10
          }
        }
      }
    },
    "niuqXoKCTAq2pNmm8cUi7g" : {
      "algorithms" : { }
    }
  }
}

```

For BWC, user still can 
1. use `GET /_plugins/_ml/stats` to get all stats 
2. use `GET /_plugins/_ml/stats/{stat}` to get specific cluster and node level stat, for example `GET /_plugins/_ml/stats/ml_model_index_status,ml_node_total_request_count`
3. use `GET /_plugins/_ml/{nodeId}/stats` to get all stats on specific node
4. use `GET /_plugins/_ml/{nodeId}/stats/{stat}` to get specific stat on specific node

### Issues Resolved
https://github.com/opensearch-project/ml-commons/issues/329
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: plugin/src/main/java/org/opensearch/ml/rest/RestStatsMLAction.java [hunk 1]
```diff
@@ -5,40 +5,66 @@
...
 import org.opensearch.rest.RestRequest;
-import org.opensearch.rest.action.RestToXContentListener;
+import org.opensearch.rest.RestStatus;
 
 import com.google.common.annotations.VisibleForTesting;
 import com.google.common.collect.ImmutableList;
 
-//TODO: add IT
+@Log4j2
 public class RestStatsMLAction extends BaseRestHandler {
```
Comment (Zhangxunmt) [inline plugin/src/main/java/org/opensearch/ml/rest/RestStatsMLAction.java:52]: nitpick: Should we rename it to RestMLStatsAction? I just realized all other rest actions have the same naming convention except this one.

File: plugin/src/main/java/org/opensearch/ml/rest/RestStatsMLAction.java [hunk 2]
```diff
@@ -58,60 +84,147 @@ public List<Route> routes() {
...
+            }
+        };
+    }
 
-        Set<String> validStats = mlStats.getStats().keySet();
-        if (isAllStatsRequested(requestedStats)) {
-            mlStatsRequest.setRetrieveAllStats(true);
+    MLStatsInput createMlStatsInputFromRequestParams(RestRequest request) {
+        MLStatsInput mlStatsInput = new MLStatsInput();
+        Optional<String[]> nodeIds = splitCommaSeparatedParam(request, "nodeId");
```
Comment (Zhangxunmt) [inline plugin/src/main/java/org/opensearch/ml/rest/RestStatsMLAction.java:128]: This is to parse all the nodeIds in the url in "GET /_plugins/_ml/{nodeId}/stats"? For example, GET /_plugins/_ml/nodeId1,nodeId2/stats would return stats for both id1 and id2?

**Review by Zhangxunmt — APPROVED** [review]

File: plugin/src/main/java/org/opensearch/ml/rest/RestStatsMLAction.java (hunk 2 above)
Comment (ylwu-amzn) [reply plugin/src/main/java/org/opensearch/ml/rest/RestStatsMLAction.java:128]: Correct, user can get multiple node's stats by separating node ids with comma.

File: plugin/src/main/java/org/opensearch/ml/rest/RestStatsMLAction.java [hunk 3]
```diff
@@ -58,60 +84,147 @@ public List<Route> routes() {
...
+        if (nodeIds.isPresent()) {
+            mlStatsInput.getNodeIds().addAll(Arrays.asList(nodeIds.get()));
+        }
+        Optional<String[]> stats = splitCommaSeparatedParam(request, "stat");
+        if (stats.isPresent()) {
+            for (String state : stats.get()) {
+                state = state.toUpperCase(Locale.ROOT);
+                // only support cluster and node level stats for bwc
```
[the remark this answers is not in the document]
Comment (ylwu-amzn) [reply plugin/src/main/java/org/opensearch/ml/rest/RestStatsMLAction.java:136]: This is to support the old REST API which doesn't support `MLStatsInput` in HTTP body. But as we put the algorithm and action level in nested structure, user can only get cluster and node level stats. To get algorithm or action level stats, user should use `MLStatsInput` like
```
GET /_plugins/_ml/stats
{
  "target_stat_levels": [
    "algorithm"
  ],
  "algorithms": [
    "kmeans","anomaly_localization"
  ]
}
```

File: plugin/src/main/java/org/opensearch/ml/rest/RestStatsMLAction.java (hunk 2 above)
Comment (jackiehanyang) [reply plugin/src/main/java/org/opensearch/ml/rest/RestStatsMLAction.java:128]: Is this going to raise any security concern? Like can a person call this on a node that they doesn't own?

**Review by jackiehanyang — COMMENTED** [review]
half way done, will finish reviewing the rest later.

File: plugin/src/main/java/org/opensearch/ml/action/stats/MLStatsNodeResponse.java [hunk 4]
```diff
@@ -29,18 +39,27 @@ public class MLStatsNodeResponse extends BaseNodeResponse implements ToXContentF
      */
     public MLStatsNodeResponse(StreamInput in) throws IOException {
         super(in);
-        this.statsMap = in.readMap(StreamInput::readString, StreamInput::readGenericValue);
+        if (in.readBoolean()) {
+            this.nodeStats = in.readMap(stream -> stream.readEnum(MLNodeLevelStat.class), StreamInput::readGenericValue);
+        }
+        if (in.readBoolean()) {
```
[the remark this answers is not in the document]
Comment (ylwu-amzn) [reply plugin/src/main/java/org/opensearch/ml/action/stats/MLStatsNodeResponse.java:45]: No, line 42 is to check if `nodeStats` exists, line45 is to check if `algorithmStats` exists.

File: plugin/src/main/java/org/opensearch/ml/action/stats/MLStatsNodeResponse.java [hunk 5]
```diff
@@ -57,21 +76,61 @@ public static MLStatsNodeResponse readStats(StreamInput in) throws IOException {
...
+            out.writeBoolean(true);
+            out.writeMap(nodeStats, (stream, v) -> stream.writeEnum(v), StreamOutput::writeGenericValue);
+        } else {
+            out.writeBoolean(false);
+        }
+        if (algorithmStats != null) {
+            out.writeBoolean(true);
+            out.writeMap(algorithmStats, (stream, v) -> stream.writeEnum(v), (stream, stats) -> stats.writeTo(stream));
+        } else {
+            out.writeBoolean(false);
```
[the remark this answers is not in the document]
Comment (ylwu-amzn) [reply plugin/src/main/java/org/opensearch/ml/action/stats/MLStatsNodeResponse.java:89]: It's a flag. We need to use different flags to indicate `nodeStats` and `algorithmStats` is null or not.

File: plugin/src/main/java/org/opensearch/ml/rest/RestStatsMLAction.java (hunk 1 above)
Comment (ylwu-amzn) [reply plugin/src/main/java/org/opensearch/ml/rest/RestStatsMLAction.java:52]: Good catch. Will rename

**Review by Zhangxunmt — APPROVED** [review]

**Pull request merged**